File: frontend/src/api/quizAttempts.ts
import { apiRequest } from './client';
import type {
  CurrentQuestionResponse,
  QuestionOption,
  QuizResult,
  StartQuizAttemptResponse,
  SubmitAnswerRequest,
  SubmitAnswerResponse,
} from '../types';

// Wire shapes actually returned by the backend (see
// backend/app/schemas/quiz_attempt.py and question.py) -- these use `id` /
// `question_index` (0-based), not the `question_id` / `question_number`
// (1-based) naming the rest of the frontend is written against. Mapped to
// the app-facing types right here at the API boundary.

interface QuestionWire {
  id: string;
  chapter_id: string;
  text: string;
  options: QuestionOption[];
  question_index: number;
  total_questions: number;
}

interface AttemptWire {
  id: string;
  total_questions: number;
}

interface StartAttemptWire {
  attempt: AttemptWire;
  current_question: QuestionWire;
}

interface ReviewItemWire {
  id: string;
  text: string;
  options: QuestionOption[];
  correct_option: string;
  selected_option: string | null;
  is_correct: boolean | null;
}

interface QuizResultWire {
  attempt_id: string;
  correct_count: number;
  total_questions: number;
  score_percent: number;
  review: ReviewItemWire[];
}

function toQuestionResponse(q: QuestionWire): CurrentQuestionResponse {
  return {
    question_id: q.id,
    text: q.text,
    options: q.options,
    question_number: q.question_index + 1,
    total_questions: q.total_questions,
  };
}

export async function startQuizAttempt(
  chapterId: string,
  userId: string,
): Promise<StartQuizAttemptResponse> {
  const data = await apiRequest<StartAttemptWire>('/quiz-attempts', {
    method: 'POST',
    body: { chapter_id: chapterId },
    userId,
  });
  return {
    attempt_id: data.attempt.id,
    total_questions: data.attempt.total_questions,
    current_question: toQuestionResponse(data.current_question),
  };
}

export async function getCurrentQuestion(
  attemptId: string,
  userId: string,
): Promise<CurrentQuestionResponse> {
  const data = await apiRequest<QuestionWire>(`/quiz-attempts/${attemptId}/current-question`, {
    userId,
  });
  return toQuestionResponse(data);
}

export function submitAnswer(
  attemptId: string,
  payload: SubmitAnswerRequest,
  userId: string,
): Promise<SubmitAnswerResponse> {
  return apiRequest<SubmitAnswerResponse>(`/quiz-attempts/${attemptId}/answers`, {
    method: 'POST',
    body: payload,
    userId,
  });
}

export async function getQuizResult(attemptId: string, userId: string): Promise<QuizResult> {
  const data = await apiRequest<QuizResultWire>(`/quiz-attempts/${attemptId}/result`, { userId });
  return {
    correct_count: data.correct_count,
    total_questions: data.total_questions,
    score_percent: data.score_percent,
    review: data.review.map((r) => ({
      question_id: r.id,
      text: r.text,
      options: r.options,
      correct_option: r.correct_option,
      selected_option: r.selected_option,
      is_correct: !!r.is_correct,
    })),
  };
}

File: backend/app/routers/quiz_attempts.py
from fastapi import APIRouter, Depends
from motor.motor_asyncio import AsyncIOMotorDatabase

from app.core.database import get_db
from app.core.security import get_current_user_id
from app.schemas.quiz_attempt import (
    AnswerResultOut,
    AnswerSubmit,
    QuizAttemptCreate,
    QuizAttemptStartOut,
    QuizResultOut,
)
from app.schemas.question import QuestionPublicOut
from app.services import quiz_service

router = APIRouter(prefix="/api/v1/quiz-attempts", tags=["quiz-attempts"])


@router.post("", response_model=QuizAttemptStartOut)
async def start_or_resume_attempt(
    payload: QuizAttemptCreate,
    user_id: str = Depends(get_current_user_id),
    db: AsyncIOMotorDatabase = Depends(get_db),
):
    result = await quiz_service.start_or_resume_attempt(db, user_id, payload.chapter_id)
    return result


@router.get("/{attempt_id}/current-question", response_model=QuestionPublicOut)
async def get_current_question(
    attempt_id: str,
    user_id: str = Depends(get_current_user_id),
    db: AsyncIOMotorDatabase = Depends(get_db),
):
    return await quiz_service.get_current_question(db, attempt_id, user_id)


@router.post("/{attempt_id}/answers", response_model=AnswerResultOut)
async def submit_answer(
    attempt_id: str,
    payload: AnswerSubmit,
    user_id: str = Depends(get_current_user_id),
    db: AsyncIOMotorDatabase = Depends(get_db),
):
    return await quiz_service.submit_answer(db, attempt_id, user_id, payload.question_id, payload.selected_option)


@router.get("/{attempt_id}/result", response_model=QuizResultOut)
async def get_result(
    attempt_id: str,
    user_id: str = Depends(get_current_user_id),
    db: AsyncIOMotorDatabase = Depends(get_db),
):
    return await quiz_service.get_result(db, attempt_id, user_id)
